js/ui/control/scale.js

```diff
@@ -46,17 +48,33 @@ function updateScale(map, scale, options) {
     // Using spherical law of cosines approximation, the real distance is
     // found between the two coordinates.
     var maxWidth = options && options.maxWidth || 100;
+    var ratio;
 
     var y = map._container.clientHeight / 2;
     var maxMeters = getDistance(map.unproject([0, y]), map.unproject([maxWidth, y]));
-
     // The real distance corresponding to 100px scale length is rounded off to
     // near pretty number and the scale length for the same is found out.
-    var meters = getRoundNum(maxMeters);
-    var ratio = meters / maxMeters;
-    scale.style.width = maxWidth * ratio + 'px';
-    scale.innerHTML = meters < 1000 ? meters + ' m' : (meters / 1000) + ' km';
-
+    // Default unit of the scale is based on User's locale.
+    if (options && options.unit === 'Imperial' || (!options.unit && window.navigator.language === 'en-US')) {
+        var maxFeet = 3.2808 * maxMeters;
+        if (maxFeet > 5280) {
+            var maxMiles = maxFeet / 5280;
+            var miles = getRoundNum(maxMiles);
+            ratio = miles / maxMiles;
+            scale.style.width = maxWidth * ratio + 'px';
+            scale.innerHTML = miles + ' mi';
+        } else {
+            var feet = getRoundNum(maxFeet);
+            ratio = feet / maxFeet;
+            scale.style.width = maxWidth * ratio + 'px';
+            scale.innerHTML = feet + ' ft';
+        }
+    } else {
+        var meters = getRoundNum(maxMeters);
+        ratio = meters / maxMeters;
+        scale.style.width = maxWidth * ratio + 'px';
+        scale.innerHTML = meters < 1000 ? meters + ' m' : (meters / 1000) + ' km';
```

Comment:
There is some very similar code in three places here. Can we reduce the repetition (e.g. by cutting out some function)?